### PR TITLE
Add contribution guide

### PR DESCRIPTION
**Pull request by laurmaedje**
Currently, the contribution process is not particularly well-defined. A contribution that is rejected after a lot of work went into it, is of course a very frustrating experience. This guide thus aims to clarify the steps a contribution should go through before being merged. It also provides a bit of background on how the process is affected by the fact that Typst is not only an open source project, but also a startup.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+   project with a long-term vision and it's frustrating to find out that your
+   idea does not align with that vision _after_ you have already implemented
+   something.
+2. Fork the Typst repository and start with your contribution. If you, at any
+   point in this process, are unsure about how to do something in the Typst
+   codebase, reach out to a maintainer or a more experienced contributor. Also
+   have a look at the [`ARCHITECTURE.md`][architecture] file. It explains how
+   the compiler works.
+3. Create a pull request (PR) in the Typst repository, outlining your
+   contribution, the technical rationale behind it, and if it is a new features,
```
Comment (PgBiel) [inline CONTRIBUTING.md:19]: Slight grammatical improvement
```suggestion
   contribution, the technical rationale behind it, and, if it is a new feature,
```

**Review by PgBiel — COMMENTED** [review]
Looking great!

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+   find/open an [issue] or start a thread in the [#contributors] channel on
+   Discord to discuss the design. Feel also free to ping a maintainer/team
+   member to get some input on your idea. Don't be shy! Typst is a complex
+   project with a long-term vision and it's frustrating to find out that your
+   idea does not align with that vision _after_ you have already implemented
+   something.
+2. Fork the Typst repository and start with your contribution. If you, at any
+   point in this process, are unsure about how to do something in the Typst
+   codebase, reach out to a maintainer or a more experienced contributor. Also
+   have a look at the [`ARCHITECTURE.md`][architecture] file. It explains how
```
Comment (YDX-2147483647) [inline CONTRIBUTING.md:16]: The definition of the link `architecture` in `CONTRIBUTING.md` is missing. Maybe add the following line (from `README.md`)?

```markdown
[architecture]: https://github.com/typst/typst/blob/main/ARCHITECTURE.md
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
+# Contributing to Typst
+Thank you for considering to contribute to Typst. We want to foster a welcoming
+and productive atmosphere for contributers. Therefore, we outline a few steps to
```
Comment (TomBinford) [inline CONTRIBUTING.md:3]:
```suggestion
and productive atmosphere for contributors. Therefore, we outline a few steps to
```

**Comment by astrale-sharp** [comment]
Very good, I think clearer rules like this will lead to a better contributing experience

**Pull request merged**